File: backend/app/schemas/admin_staff.py
"""Admin staff management schemas (ADMIN-08B)."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.core.staff_admin_constants import (
    STAFF_ADMIN_ACTION_LIST_PAGE_SIZE_MAX,
    STAFF_ADMIN_LIST_PAGE_SIZE_MAX,
    STAFF_ADMIN_REASON_MAX_LENGTH,
)


class AdminStaffListItem(BaseModel):
    id: UUID
    email: str
    full_name: str
    is_active: bool
    roles: list[str]
    permissions: list[str]
    created_at: datetime
    updated_at: datetime


class AdminStaffListResponse(BaseModel):
    items: list[AdminStaffListItem]
    total: int = Field(ge=0)
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=STAFF_ADMIN_LIST_PAGE_SIZE_MAX)


class AdminStaffAdminSummaryResponse(BaseModel):
    """Aggregated platform access metrics (STAFF-V10-01)."""

    model_config = ConfigDict(extra="forbid")

    generated_at: datetime
    total: int = Field(ge=0)
    active: int = Field(ge=0)
    suspended: int = Field(ge=0)
    super_admins: int = Field(ge=0)
    city_admins: int = Field(ge=0)
    moderators: int = Field(ge=0)
    dominant_role: str | None = None


class AdminStaffDetailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    email: str
    full_name: str
    city: str | None = None
    is_active: bool
    is_verified: bool
    roles: list[str]
    permissions: list[str]
    created_at: datetime
    updated_at: datetime


class AdminStaffAssignRoleRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    role: str = Field(min_length=1, max_length=64)
    reason: str | None = Field(default=None, max_length=STAFF_ADMIN_REASON_MAX_LENGTH)


class AdminStaffReasonRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reason: str | None = Field(default=None, max_length=STAFF_ADMIN_REASON_MAX_LENGTH)


class AdminStaffActionActorSummary(BaseModel):
    id: UUID
    email: str
    display_name: str | None = None


class AdminStaffActionListItem(BaseModel):
    action: str
    previous_roles: list[str] | None = None
    new_roles: list[str] | None = None
    reason: str | None = None
    actor_user: AdminStaffActionActorSummary | None = None
    created_at: datetime


class AdminStaffActionListResponse(BaseModel):
    items: list[AdminStaffActionListItem]
    total: int = Field(ge=0)
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=STAFF_ADMIN_ACTION_LIST_PAGE_SIZE_MAX)

File: frontend/packages/types/src/admin-staff.ts
/** Admin platform staff types (ADMIN-08C). */

import type { RoleKey } from "./auth";

export type AdminStaffPlatformRole = Extract<
  RoleKey,
  "SUPER_ADMIN" | "CITY_ADMIN" | "MODERATOR"
>;

export type AdminStaffActionType =
  | "assign_role"
  | "revoke_role"
  | "suspend"
  | "reactivate";

export interface AdminStaffListItem {
  id: string;
  email: string;
  full_name: string;
  is_active: boolean;
  roles: string[];
  permissions: string[];
  created_at: string;
  updated_at: string;
}

export interface AdminStaffListResponse {
  items: AdminStaffListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface AdminStaffListParams {
  role?: string;
  is_active?: boolean;
  page?: number;
  page_size?: number;
}

export interface AdminStaffDetailResponse {
  id: string;
  email: string;
  full_name: string;
  city: string | null;
  is_active: boolean;
  is_verified: boolean;
  roles: string[];
  permissions: string[];
  created_at: string;
  updated_at: string;
}

export interface AdminStaffActionActor {
  id: string;
  email: string;
  display_name: string | null;
}

export interface AdminStaffActionItem {
  action: AdminStaffActionType | string;
  previous_roles: string[] | null;
  new_roles: string[] | null;
  reason: string | null;
  actor_user: AdminStaffActionActor | null;
  created_at: string;
}

export interface AdminStaffActionListResponse {
  items: AdminStaffActionItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface AdminStaffActionListParams {
  page?: number;
  page_size?: number;
}
